Share LTPA keys

kabicin opened the pull request:
PRs for #530, #531, https://github.com/WASdev/websphere-liberty-operator/issues/532


**Shared LTPA key scenario:** 
Suppose you have a WebSphereLibertyApplication called `example`
The `example-ltpa-secret` stores the unencrypted `password` and also the `lastRotation` for the LTPA keys file.
If it is not found, the operator will delete any existing Liberty Deployment/StatefulSet and create the Secret by generating a random string of length 15, sets the current time as the last rotation.
Just after creation of the Secret, it starts two Jobs to
1. Generate the LTPA key
2. Determines the encrypted password from the random string of length 15.

After this happens, the operator will generate a ConfigMap representing the `ltpa.xml` file populating the field with the encrypted password generated from the second Job. The ltpa.xml file is stored in `/config/ltpa/xml/ltpa.xml`.

It might take some time to get the output from the Pod logs. So only when the ConfigMap is successfully created, the two Jobs from earlier are then deleted.

When the Liberty pod(s) `example` eventually start (later in the reconcile), an `initContainer` runs. It's sole purpose is to copy the `/config/ltpa/xml/ltpa.xml` file into the `/config/configDropins/overrides` folder so that Liberty can recognize the LTPA flag to prevent from generating a new one itself.

leochr commented on utils/utils.go:
```diff
@@ -585,3 +590,141 @@ func GetWLOLicenseAnnotations() map[string]string {
+		pts.Spec.Volumes = append(pts.Spec.Volumes, vol)
+	}
+
+	// Create an initContainer to copy /config/ltpa/xml/ltpa.xml into the emptyDir volume at /config/configDropins/overrides
```
@kabicin please try the `subPath` approach (mentioned in the articles [here](https://imantumorang.com/posts/mount-file-to-kubernetes-pod-without-deleting-the-existing-file-in-the-docker-container-in-the-same-directory/) and [here](https://www.golinuxcloud.com/mount-configmap-as-file-in-existing-directory/)) to include the LTPA config into an existing directory with files. Assuming it achieves the same result, I think this approach would be better than using the init container (the primary/app container is not started until init containers start and complete their process. Not a significant delay, but still some delay):

kabicin replied:
I was able to use subPath so I removed the initContainer

leochr commented on controllers/ltpa_keys_sharing.go:
```diff
@@ -0,0 +1,315 @@
+	}
+
+	// Create the Liberty Server XML ConfigMap if it doesn't exist
+	configMapErr := r.GetClient().Get(context.TODO(), types.NamespacedName{Name: ltpaServerXMLConfigMap.Name, Namespace: ltpaServerXMLConfigMap.Namespace}, ltpaServerXMLConfigMap)
```
Since the encrypted password is included in the config, perhaps it's better to use Secret for server config as well

kabicin replied:
I changed this to a Secret now

leochr commented:
@kabicin Received this error when I tried with the CatalogSource image for the branch:

<img width="1344" alt="ltpa_error_1" src="https://github.com/WASdev/websphere-liberty-operator/assets/18424008/19bbc1ff-22b2-4780-be5a-55ccb1a09d77">

The necessary RBAC permissions for the LTPA Job should also be added to the Operator's RBAC/serviceaccount

leochr commented:
@kabicin Ran into another RBAC error, this time for Role and RoleBindings:

```
reflector.go:324] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: failed to list *v1.Role: roles.rbac.authorization.k8s.io is forbidden: User "system:serviceaccount:ltpa:wlo-controller-manager" cannot list resource "roles" in API group "rbac.authorization.k8s.io" in the namespace "ltpa"
E1026 14:47:22.687286 1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.2/tools/cache/reflector.go:167: Failed to watch *v1.Role: failed to list *v1.Role: roles.rbac.authorization.k8s.io is forbidden: User "system:serviceaccount:ltpa:wlo-controller-manager" cannot list resource "roles" in API group "rbac.authorization.k8s.io" in the namespace "ltpa"
```

Added the RBAC for them manually, but then I got this message. I believe it's because `create_ltpa_keys.sh` is not included with the operator image. Please add it under the `controllers/assets` directory in both GitHub and  operator image - similar to WSA - [here](https://github.ibm.com/websphere/automation-operator/blob/9b2c529c644faaeca14b4e6a9d89c710b343eb5f/Dockerfile#L57).
```
ERROR controllers.WebSphereLibertyApplication Failed to generate the shared LTPA Keys file {"Request.Namespace": "ltpa", "Request.Name": "liberty-app", "error": "open utils/create_ltpa_keys.sh: no such file or directory"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
```

kabicin commented:
@leochr Thanks, I've added the RBAC perms and moved the script under `controllers/assets` now.

leochr commented:
@kabicin Thanks for the RBAC updates.

Suggestions for some name changes (to use wlo-managed consistently):
- wlo-ltpa-server-xml --> wlo-managed-ltpa-server-xml
- /config/ltpa --> /config/managedLTPA
- /config/configDropins/overrides/ltpa.xml --> /config/configDropins/overrides/managedLTPA.xml


wlo-ltpa-server-xml is owned by the WebSphereLibertyApplication instance, but as it's shared among all instances, it shouldn’t be.

Please add the default set of labels (including `managed-by`) to new resources (secrets, serviceaccount, role, rolebinding, etc). Thanks

kabicin commented:
@leochr I removed the ownerReference on the LTPA XML Secret and prevented it from being deleted when the app turns manageLTPA off. Also updated naming in the last commit.

leochr commented:
@kabicin Suggestions for the status messages:
- “Waiting for liberty-app9 to generate the shared LTPA keys file for the ltpa namespace.” 
--> 
“Waiting for WebSphereLibertyApplication instance ‘liberty-app9’ to generate the shared LTPA keys file for the namespace ‘ltpa’.”
- “Waiting for Job wlo-managed-ltpa-keys-generation to be completed.” --> “Waiting for the LTPA key to be generated by Job ‘wlo-managed-ltpa-keys-generation’.”

One scenario to improve:
- create the first WebSphereLibertyApplication instance in the namespace with an invalid image
- the Job will be kicked off. The pod fails with ImagePullBackOff. Job never completes.
- Update the image in the CR to specify a valid image
- The Job still uses the invalid image and holts the reconcile process for this CR instance as well as all other CR instances with manageLTPA set to true

kabicin commented:
@leochr 
- Added new status message changes
- Renamed `wlo-ltpa-job-request` to `wlo-managed-ltpa-job-request`.
- Added a step in the app image change detection, to restart the LTPA generation process (only if the LTPA Secret was never generated successfully from the previous app image). The "LTPA key generation leader" which is the app instance, designated to create/delete the LTPA related resources, will be the one listening for this change.

leochr commented:
@kabicin Thanks for the updates. Just want to make sure that the Job won't be deleted before it has the chance to complete (with a valid image).

leochr commented on utils/utils.go:
```diff
@@ -585,3 +591,123 @@ func GetWLOLicenseAnnotations() map[string]string {
+
+func CustomizeLTPAServerXML(xmlSecret *corev1.Secret, la *wlv1.WebSphereLibertyApplication, encryptedPassword string) {
+	xmlSecret.StringData = make(map[string]string)
+	keysFileName := strings.Replace(ltpaKeysMountPath, "/config", "${server.config.dir}", 1)
```
Since the file is mounted to `ltpaKeysMountPath`, it's probably better to use that location as is.

kabicin replied:
For some reason the Liberty implementation for parsing `<ltpa keysFileName="?"></ltpa>` won't do a deep folder traversal for the symlink `/config`, so I had to change it to `${server.config.dir}` for it to read the file properly.

leochr commented:
@kabicin Nvm. I see that the Job deletion only happens if the image changed (and LTPA secret hasn't been created yet). So it should be fine I think.

leochr commented:
@kabicin I'll go ahead and merge this PR, so we can get a build going. The rest of the code review and updates can be made via separate PR(s).
